README
$Id: README 105 2008-04-07 06:00:07Z phf $

===================
DASM 2.20.11 README
===================

    Welcome to DASM, a versatile macro assembler with support for several
    8-bit microprocessors including MOS 6502 & 6507; Motorola 6803, 68705,
    and 68HC11; Hitachi HD6303 (extended Motorola 6801); and Fairchild F8.

    This file describes the DASM distribution, how to compile and install
    DASM, and where to get more information about DASM.

    DASM's homepage is http://dasm-dillon.sourceforge.net/ since release
    2.20.11.

----------------
The Distribution
----------------

    The DASM distribution contains the following important files and
    directories of interest to all users:

        bin/        Executables for DASM and related tools
                    (only if you downloaded a binary distribution)
        doc/        Documentation and manuals
        machines/   Support files for various 8-bit machines
        AUTHORS     List of contributors to DASM
        COPYING     GNU General Public License
        FUTURE      Plans for the next releases of DASM
        NEWS        Recent changes to DASM
        README      The file you are reading right now :-)

    In addition, developers are going to be interested in the following
    files and directories:

        src/        Source code for DASM and related tools
        test/       Test cases for DASM and related tools
                    (as of release 2.20.11 the test framework is incomplete)
        ChangeLog   Source level changes to DASM
        Makefile    Makefile to build DASM executables, run tests, and
                    create distributions

    Even more files for developers are available from the DASM Subversion
    repository, see http://dasm-dillon.sf.net/ for details.

------------------------
Compiling and Installing
------------------------

    If you are using DASM on a Unix system, you should be able to simply
    give the command

        make

    in the root directory of the DASM distribution and DASM should build.
    A bin/ directory containing DASM executables will be created as part
    of this process. You can also give the command

        make test

    to run all the test cases that come with the distribution. Note that
    as of release 2.20.11 the test framework is incomplete and probably
    only remotely comprehensible for developers. :-/ The often-used

        make install

    does not work as of release 2.20.11 either, mostly because DASM does
    not yet use GNU autotools. You have to install DASM by hand, either
    by adding the bin/ directory to your path, or by copying executables
    to another directory in your path. For include files in the machines/
    directory you currently have to use DASM's -I command line option.

    NOTE: Several compiler warnings will be displayed during the build
          process. We are confident that you can ignore these warnings.
          They will be dealt with in a future release. (And if you know
          how to fix them reliably, we welcome your contributions!)

    DASM has been built and tested successfully on the following machines
    and operating systems:

        - SunOS 5.10 Generic_118822-11 sun4u sparc
        - Linux 2.6.23.15-80.fc7 i386 GNU/Linux
        - Darwin 8.11.0 powerpc
        - NetBSD 3.0 qube mips
        - Linux 2.6.24-1-686 i686 GNU/Linux
        - Linux-based cross-compile for Windows/Wine

    If you have successfully built and tested DASM on a different machine
    or operating system, we would love to hear about it.

----------
Using DASM
----------

    The simplest way to get a brief introduction to DASM is to run the
    bin/dasm executable without options, which will print a short help
    message summarizing all available options.

    The DASM distribution includes documentation in the doc/ directory.
    However, the documentation is incomplete as of release 2.20.11. We
    are urgently looking for volunteers to bring it up to date.

    Documentation for using DASM to produce code for the F8 processor
    is currently in the directory machines/channel-f/ but will be
    integrated into the main documentation in the future.

    The sourceforge.net project page provides access to two mailing
    lists, one for announcements from the developers and one for DASM
    users to discuss problems, help each other, and communicate with
    the developers. See http://sourceforge.net/projects/dasm-dillon/
    for details.

    There is also a vibrant community of developers writing games and
    demos for the Atari 2600 VCS using DASM. The mailing lists for the
    Stella emulator are a particularly useful resource for DASM users,
    not only those intent on programming for the VCS.

--------
Legalese
--------

    the DASM macro assembler (aka small systems cross assembler)

    Copyright (c) 1988-2002 by Matthew Dillon.
    Copyright (c) 1995 by Olaf "Rhialto" Seibert.
    Copyright (c) 2003-2008 by Andrew Davie.
    Copyright (c) 2008 by Peter H. Froehlich.

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.
 
    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License along
    with this program; if not, write to the Free Software Foundation, Inc.,
    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.



NEWS
$Id: NEWS 105 2008-04-07 06:00:07Z phf $

This file records only user-visible changes to DASM, the ChangeLog
file records all internal changes as well (almost all anyway).

=============
The DASM News
=============

2.20.11 -- 07 April 2008

    * Peter H. Froehlich <phf at acm dot org> took over as maintainer
      from Andrew Davie, but both Andrew and Thomas Mathys still help.

    * DASM moved to sourceforge.net, the new website for DASM is now
      available at <http://dasm-dillon.sourceforge.net/>.

    * Fixed (the obvious version of) the long-standing "-1 bug" that
      made DASM reject sources with negative integer literals; there
      may still be signed/unsigned problems in other areas; there is
      certainly still a range checking error.

    * Fixed the usage message to reflect that symbol table sorting is
      controlled by -T and not by -t as the message claimed; also the
      format of the usage message is now more GNU-like.

    * The range of values for -T is now checked, you have to stick to
      0 or 1, arbitrary numbers will not work anymore.

    * Removed spurious output when running DASM. In general command
      line tools should only make output when there's something to
      report (a central Unix tenent). Sorry if you miss the messages,
      just look at the usage information instead. :-)

    * Added new -E# option to select format of error messages; -E0 is
      the default MS style, -E1 is "classic" style from Matthew Dillon,
      -E2 is GNU style from their coding standards.

2.20.10 -- 31 October 2004

    * Thomas Mathys added support for the Fairchild F8 processor to
      DASM. Thank you, Thomas!

    * Support files for the Channel-F Video Entertainment System (VES)
      added.

2.20.08 -- unreleased

    * Documentation changed from Word documents back to text files.

2.20.07 -- 17 July 2003

    * No user-visible changes.

2.20.05 -- unreleased

    * Symbol table dump is now performed even if an error occurred.

    * Symbol table dump now displays string symbols in string form.

2.20.04 -- 24 April 2003

    * Command line options now allow "/" as prefix in addition to "-"
      as before.

    * Division by zero is now treated as an unrecoverable error.

    * Error messages now contain more information about the error,
      including the actual source code where possible.

    * Output reworked. Typically, the -v option is now unnecessary as
      more appropriate messages allow you to find errors better.
      Compatibility with the -v option has been maintained, although
      there may be a few minor changes.

    * Branch out of range errors now list the distance to the branch.

    * Undefined symbols are now always listed in a separate table. If
      an unresolved symbol causes an error, the table is displayed
      automatically at the end of assembly. (Note: It can be alright
      to have a referenced but undefined symbol, for example when
      IFCONST is used. In this case, the undefined table will only
      be displayed if there is another undefined label, but it will
      contain *all* unresolved symbols.)

    * New command line option -T# where # is 0 or 1; 0 sorts symbol
      table alphabetically (default), 1 sorts symbol table by address.
      If available memory precludes sorting, the table is unsorted.

2.20.02 -- 24 April 2003

    * John Saeger contributed support for all illegal 6502 opcodes.
      Thank you, John! This fixes the LAX opcode as well.

2.20.01 -- 22 March 2003

    * Andrew Davie <atari2600 at taswegian dot com> took up the DASM
      project, starting from Olaf "Rhialto" Seibert's 2.12.04 version
      with both Olaf's and Matthew Dillon's blessing.

    * REPEAT now detects negative loop count (used to lock up).

    * Errors are now in the format "file (line): Error: string" to
      be compatible with MS VisualStudio.

    * Support files for the Atari 2600 Video Computer System (VCS)
      added.

    * Support for illegal 6502/6507 opcodes NOP zp and LAZ (zp),y
      added.

2.12.14 -- 19 August 1995

    * Olaf "Rhialto" Seibert's release of DASM with many changes to
      Matthew Dillon's 2.12 release.

Note that everything before 2.20.11 was *reconstructed* by Peter
Froehlich for the 2.20.11 release. Peter had no first-hand knowledge
of the early years but tried to be as accurate as possible, working
from multiple sources. Please feel free to submit any corrections you
may have regarding dates, features, and people.



src/globals.c
/*
    $Id: globals.c 72 2008-04-05 20:49:29Z phf $

    the DASM macro assembler (aka small systems cross assembler)

    Copyright (c) 1988-2002 by Matthew Dillon.
    Copyright (c) 1995 by Olaf "Rhialto" Seibert.
    Copyright (c) 2003-2008 by Andrew Davie.
    Copyright (c) 2008 by Peter H. Froehlich.

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License along
    with this program; if not, write to the Free Software Foundation, Inc.,
    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
*/

/*
 *  GLOBALS.C
 */

#include "asm.h"

SVNTAG("$Id: globals.c 72 2008-04-05 20:49:29Z phf $");

SYMBOL *SHash[SHASHSIZE];   /*	symbol hash table   */


MNEMONIC    *MHash[MHASHSIZE];   /*	mnemonic hash table */
INCFILE *pIncfile;	    /*	include file stack  */
REPLOOP *Reploop;	    /*	repeat loop stack   */
SEGMENT *Seglist;	    /*	segment list	    */
SEGMENT *Csegment;	    /*	current segment     */
IFSTACK *Ifstack;	    /*	IF/ELSE/ENDIF stack */
char	*Av[256];	    /*	up to 256 arguments */
char	Avbuf[512];
unsigned char	MsbOrder = 1;
int	Mnext;
char	Inclevel;
unsigned int	Mlevel;
unsigned long	Localindex;	   /*  to generate local variables */
unsigned long	Lastlocalindex;

unsigned long	Localdollarindex;
unsigned long	Lastlocaldollarindex;

unsigned long Processor = 0;
bool bTrace = false;
bool Xdebug;

unsigned char	Outputformat;

unsigned long   Redo_why = 0;
int	Redo_eval = 0;	   /*  infinite loop detection only    */
int Redo = 0;


unsigned long	Redo_if = 0;

char	ListMode = 1;
unsigned long	CheckSum;	    /*	output data checksum		*/

int F_format = FORMAT_DEFAULT;

/* -T option [phf] */
sortmode_t F_sortmode = SORTMODE_DEFAULT;
/* -E option [phf] */
errorformat_t F_errorformat = ERRORFORMAT_DEFAULT;

unsigned char	 F_verbose;
const char	*F_outfile = "a.out";
char	*F_listfile;
char	*F_symfile;
FILE	*FI_listfile;
FILE	*FI_temp;
unsigned char	 Fisclear;
unsigned long	 Plab, Pflags;

/*unsigned int	Adrbytes[]  = { 1, 2, 3, 2, 2, 2, 3, 3, 3, 2, 2, 2, 3, 1, 1, 2, 3 };*/
unsigned int	Cvt[]	    = { 0, 2, 0, 6, 7, 8, 9, 0, 0, 0, 0, 0, 0, 4, 5, 0, 0 };
unsigned int	Opsize[]    = { 0, 1, 2, 1, 1, 1, 2, 2, 2, 2, 1, 1, 2, 0, 0, 1, 1 };

MNEMONIC Ops[] = {
    { NULL, v_list    , "list",           0,      0, {0,} },
    { NULL, v_include , "include",        0,      0, {0,} },
    { NULL, v_seg     , "seg",            0,      0, {0,} },
    { NULL, v_hex     , "hex",            0,      0, {0,} },
    { NULL, v_err     , "err",            0,      0, {0,} },
    { NULL, v_dc      , "dc",             0,      0, {0,} },
    { NULL, v_dc      , "byte",           0,      0, {0,} },
    { NULL, v_dc      , "word",           0,      0, {0,} },
    { NULL, v_dc      , "long",           0,      0, {0,} },
    { NULL, v_ds      , "ds",             0,      0, {0,} },
    { NULL, v_dc      , "dv",             0,      0, {0,} },
    { NULL, v_end     , "end",            0,      0, {0,} },
    { NULL, v_trace   , "trace",          0,      0, {0,} },
    { NULL, v_org     , "org",            0,      0, {0,} },
    { NULL, v_rorg    , "rorg",           0,      0, {0,} },
    { NULL, v_rend    , "rend",           0,      0, {0,} },
    { NULL, v_align   , "align",          0,      0, {0,} },
    { NULL, v_subroutine, "subroutine",   0,      0, {0,} },
    { NULL, v_equ     , "equ",            0,      0, {0,} },
    { NULL, v_equ     , "=",              0,      0, {0,} },
    { NULL, v_eqm     , "eqm",            0,      0, {0,} },
    { NULL, v_set     , "set",            0,      0, {0,} },
    { NULL, v_macro   , "mac",            MF_IF,  0, {0,} },
    { NULL, v_endm    , "endm",           MF_ENDM,0, {0,} },
    { NULL, v_mexit   , "mexit",          0,      0, {0,} },
    { NULL, v_ifconst , "ifconst",        MF_IF,  0, {0,} },
    { NULL, v_ifnconst, "ifnconst",       MF_IF,  0, {0,} },
    { NULL, v_if      , "if",             MF_IF,  0, {0,} },
    { NULL, v_else    , "else",           MF_IF,  0, {0,} },
    { NULL, v_endif   , "endif",          MF_IF,  0, {0,} },
    { NULL, v_endif   , "eif",            MF_IF,  0, {0,} },
    { NULL, v_repeat  , "repeat",         MF_IF,  0, {0,} },
    { NULL, v_repend  , "repend",         MF_IF,  0, {0,} },
    { NULL, v_echo    , "echo",           0,      0, {0,} },
    { NULL, v_processor,"processor",      0,      0, {0,} },
    { NULL, v_incbin  , "incbin",         0,      0, {0,} },
    { NULL, v_incdir  , "incdir",         0,      0, {0,} },
    MNEMONIC_NULL
};

